### PR TITLE
Diff

### PR DESCRIPTION
**Pull request by aarondr77**
# Changes

- [x] I ran all the tests to make sure I didn't break anything.

### CODE REVIEW

**Comment by aarondr77** [comment]
@naterush Data type is not completely formed yet, but I was thinking about returning a **list** of the following objects {sheet: "sheet1", changeType: "none/modified/inserted/deleted", changes: [{
      sheet: "sheet1", 
      cell: "cell", 
      initalElement: "55", 
      finalElement: "60"
}, ...]

**Comment by naterush** [comment]
@aarondr77 I don't think that parses, but I interpreted it as this:

```
[
        {
            sheet: "sheet1", 
            changeType: "none/modified/inserted/deleted", 
            changes: [{
                sheet: "sheet1",
                cell: "cell",
                initalElement: "55",
                finalElement: "60"
            }]
        }
    ]
```

**Comment by aarondr77** [comment]
Yes. Exactly.

**Comment by naterush** [comment]
@aarondr77 we use the variable name "sheetName" elsewhere to refer to the name. Can we move to that? And use that in merge to, if we don't already. Gotta stay strongly typed!

**Comment by naterush** [comment]
Also, I vote for initialValue and finalValue or initialElement and finalElement. An element is a thing in a list, a value makes more sense here... i think. Let me know if you disagree.

initialElement is also spelled wrong above. No worries.

File: src/constants/index.js [hunk 1]
```diff
@@ -68,4 +68,13 @@ export const headerSize = {
     LARGE: "large"
 }
 
+/*
+    The type of change to a sheet
```
Comment (naterush) [inline src/constants/index.js]: The type of change to a sheet in a difference calculation

^ maybe a better comment.

File: src/saga/diff.js [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+    firstCommitRange.load("values")
+    await context.sync();
+
+    const initialCommit = firstCommitRange.values
+    const finalCommit = await project.getCommitIDFromBranch("master");
+
+    const changes = await diff(context, initialCommit, finalCommit);
+
+    // TODO: Update last time user caught up to now
+    return changes
```
Comment (naterush) [inline src/saga/diff.js]: ;

File: src/saga/diff.js [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+        })
+    }
+
+    console.log("found the following changes", sheetChanges)
+    return sheetChanges
+}
+
+async function catchUp(context) {
+    // TODO: Find last time use caught up
+    const project = new Project(context)
```
Comment (naterush) [inline src/saga/diff.js]: ;

File: src/saga/diff.js [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+                });
+            }   
+        }
+    }
+    return changes
+}
+
+// Find all of the differences between two lists
+function rowDiff (initialRow, finalRow, sheetName, rowIndex) {
+    let changes = []
```
Comment (naterush) [inline src/saga/diff.js]: ;

File: src/saga/diff.js [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+
+    // if neither row exists, return
+    if (initialRow === undefined && finalRow === undefined) {
+        return changes
```
Comment (naterush) [inline src/saga/diff.js]: ; everywhere, plz - we should turn on a linting rule! Making an issue.

File: src/saga/diff.js [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+    }
+
+    // if only one row exists
+    if (initialRow === undefined || finalRow === undefined) {
+        return initialRow === undefined ? handleUndefinedRow(finalRow, sheetName, rowIndex, false) : handleUndefinedRow(initialRow, sheetName, rowIndex, true)
```
Comment (naterush) [inline src/saga/diff.js]: This might be easier to parse as just two ifs. Your call.

File: src/saga/diff.js [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+
+    // iterate through the rows to find changes
+    const maxLength = Math.max(initialRow.length, finalRow.length);
+    for (var i = 0; i < maxLength; i++) {
+        var initialElement = initialRow[i];
+        var finalElement = finalRow[i];
+
+        // handle if the element is undefined
+        if (initialElement === undefined) {
+            initialElement = ""
```
Comment (naterush) [inline src/saga/diff.js]: ; here and below

File: src/saga/diffUtils.js [hunk 8]
```diff
@@ -0,0 +1,43 @@
+// Find sheets that are in the final Sheet set, but not the initial Sheet set
+export function findInsertedSheets(initialSheetNames, finalSheetNames) {
+    const insertedSheetNames = finalSheetNames.filter(sheetName => {
+        return !(initialSheetNames.includes(sheetName))
+    });
+    return insertedSheetNames
```
Comment (naterush) [inline src/saga/diffUtils.js]: ; on all the returns, por favor

**Review by naterush — CHANGES_REQUESTED** [review]
This looks great. I have a few very minor nit-picks around style, but generally seems pretty good to me. Make the changes that seem reasonable, then LGTM!

**Pull request merged**